### PR TITLE
know how to propose for the current slot

### LINKED ISSUE #1753: VRF checking starts in the next slot, not the current slot
This means we're always missing a slot! https://github.com/CodaProtocol/coda/blob/ca93bb5bd5a0edbeeea98b9ddc78d7c55d566298/src/lib/consensus/proof_of_stake.ml#L1362 there shouldn't be a call to `succ` here.

### PR DESCRIPTION
**Pull request by cmr**
Closes #1753

### CODE REVIEW

**Comment by cmr** [comment]
We don't have any proposer tests. Any ideas on how to write a test for this one?

File: src/lib/consensus/proof_of_stake.ml [hunk 1]
```diff
@@ -1375,20 +1375,22 @@ let next_proposal now (state : Consensus_state.value) ~local_state ~keypair
         | None -> find_winning_slot (Epoch.Slot.succ slot)
         | Some data -> Some (slot, data)
     in
-    find_winning_slot (Epoch.Slot.succ slot)
+    find_winning_slot slot
```
Comment (nholland94) [inline src/lib/consensus/proof_of_stake.ml]: we should track whether or not we have already proposed for this slot or not yet. You can probably stick that information into local state by keeping the last checked epoch and slot.

File: src/protocols/coda_pow.ml [hunk 2]
```diff
@@ -1217,7 +1217,9 @@ module type Consensus_mechanism_intf = sig
     -> local_state:Local_state.t
     -> keypair:keypair
     -> logger:Logger.t
-    -> [`Check_again of Int64.t | `Propose of Int64.t * Proposal_data.t]
+    -> [ `Check_again of Int64.t
+       | `Propose_now of Int64.t * Proposal_data.t
+       | `Propose of Int64.t * Proposal_data.t ]
```
Comment (nholland94) [inline src/protocols/coda_pow.ml]: We should probably refer to `Unix_timestamp.t` here. It's equivalent, but the module name was added to be specific about what the `Int64.t` represented.

File: src/lib/proposer/proposer.ml [hunk 3]
```diff
@@ -367,6 +367,11 @@ module Make (Inputs : Inputs_intf) :
                   | `Check_again time ->
                       Singleton_scheduler.schedule scheduler (time_of_ms time)
                         ~f:check_for_proposal
+                  | `Propose_now (time, data) ->
+                      Singleton_supervisor.dispatch proposal_supervisor data
+                      |> ignore ;
+                      Singleton_scheduler.schedule scheduler (time_of_ms time)
```
Comment (nholland94) [inline src/lib/proposer/proposer.ml]: I don't know if it makes sense to handle time like this. It's a little confusing. If you track the most recently checked slot in local state or something, then you don't need to handle this and can just immediately call back into `next_proposal`.

Comment (cmr) [reply src/lib/proposer/proposer.ml]: How should we prevent a busy loop?

**Pull request merged**